providers/common/kdfs/kbkdf.c

```diff
@@ -0,0 +1,303 @@
+static OSSL_OP_kdf_derive_fn kbkdf_derive;
+static OSSL_OP_kdf_settable_ctx_params_fn kbkdf_settable_ctx_params;
+static OSSL_OP_kdf_set_ctx_params_fn kbkdf_set_ctx_params;
+
+static void *kbkdf_new(void *provctx)
+{
+    KBKDF *ctx;
+
+    ctx = OPENSSL_zalloc(sizeof(*ctx));
+    if (ctx == NULL) {
+        ERR_raise(ERR_LIB_PROV, ERR_R_MALLOC_FAILURE);
+        return NULL;
+    }
+
+    ctx->provctx = provctx;
+    return ctx;
+}
+
+
+static void kbkdf_free(void *vctx)
+{
+    KBKDF *ctx = vctx;
+
+    kbkdf_reset(ctx);
+    OPENSSL_free(ctx);
+}
+
+static void kbkdf_reset(void *vctx)
+{
+    KBKDF *ctx = vctx;
+
+    ossl_prov_digest_reset(&ctx->digest);
+    OPENSSL_clear_free(ctx->Context, ctx->Context_len);
+    OPENSSL_clear_free(ctx->Label, ctx->Label_len);
+    OPENSSL_clear_free(ctx->KI, ctx->KI_len);
+    memset(ctx, 0, sizeof(*ctx));
+}
+
+/* SP800-108 section 5.1. */
+static int kbkdf_derive_counter(EVP_MAC_CTX *ctx_init,
+                                unsigned char *Label, size_t Label_len,
+                                unsigned char *Context, size_t Context_len,
+                                unsigned char *K_i, size_t h, uint32_t L,
+                                unsigned char *KO, size_t KO_len)
+{
+    int ret = 0;
+    EVP_MAC_CTX *ctx = NULL;
+    size_t written = 0, to_write;
+    const unsigned char zero = 0;
+    uint32_t counter, i;
+
+    for (counter = 1; written < KO_len; counter++) {
+        i = htobe32(counter);
+
+        ctx = EVP_MAC_CTX_dup(ctx_init);
+        if (ctx == NULL)
+            goto done;
+
+        if (!EVP_MAC_update(ctx, (unsigned char *)&i, 4) ||
+            !EVP_MAC_update(ctx, Label, Label_len) ||
+            !EVP_MAC_update(ctx, &zero, 1) ||
+            !EVP_MAC_update(ctx, Context, Context_len) ||
+            !EVP_MAC_update(ctx, (unsigned char *)&L, 4))
+            goto done;
+
+        if (!EVP_MAC_final(ctx, K_i, NULL, h))
+            goto done;
+
+        to_write = KO_len - written;
+        memcpy(KO + written, K_i, MIN(to_write, h));
+        written += h;
+
+        EVP_MAC_CTX_free(ctx);
+        ctx = NULL;
+    }
+
+    ret = 1;
+done:
+    EVP_MAC_CTX_free(ctx);
+    return ret;
+}
+
+static int kbkdf_derive(void *vctx, unsigned char *key, size_t keylen)
+{
+    KBKDF *ctx = vctx;
+    OPENSSL_CTX *provctx = PROV_LIBRARY_CONTEXT_OF(ctx->provctx);
+    int ret = 0;
+    const EVP_MD *md;
+    EVP_MAC *mac;
+    const char *mdname;
+    EVP_MAC_CTX *ctx_init;
+    OSSL_PARAM params[3];
+    unsigned char *K_i = NULL;
+    uint32_t L = htobe32(keylen * 8);
+    size_t h;
+
+    /* Label and Context are permitted to be empty. Check everything else. */
+    md = ossl_prov_digest_md(&ctx->digest);
+    if (md == NULL) {
+        ERR_raise(ERR_LIB_PROV, PROV_R_MISSING_MESSAGE_DIGEST);
+        return 0;
+    } else if (ctx->KI_len == 0 || ctx->KI == NULL) {
+        ERR_raise(ERR_LIB_PROV, PROV_R_NO_KEY_SET);
+        return 0;
+    } 
+    mdname = EVP_MD_name(md);
```

Comment:
Remove all lines from 163 to here.  You can move the check of `KI` and `KI_len` down after setting the params of the MAC, 'cause at that point, you know you have a MAC to play with.